src/commonMain/kotlin/sim/SimParticipant.kt

```diff
@@ -454,35 +458,56 @@ class SimParticipant(val character: Character, val rotation: Rotation, val sim:
     fun addResource(amount: Int, type: Resource.Type, abilityName: String) {
         if(amount == 0) return
 
-        if(resource.type == type) {
-            resource.add(amount)
+        var res = resource[type]
 
-            logEvent(Event(
-                eventType = Event.Type.RESOURCE_CHANGED,
-                amount = resource.currentAmount.toDouble(),
-                delta = amount.toDouble(),
-                amountPct = resource.currentAmount / resource.maxAmount.toDouble() * 100.0,
-                abilityName = abilityName
-            ))
-        } else {
-            logger.debug { "Attempted to add resource type $type but subject resource is ${resource.type}" }
+        if(res == null) {
+            logger.debug { "Attempted to add resource type $type that is not present for participant" }
+            return
         }
+
+        res.add(amount)
+
+        logEvent(Event(
+            eventType = Event.Type.RESOURCE_CHANGED,
+            amount = res.currentAmount.toDouble(),
+            delta = amount.toDouble(),
+            amountPct = res.currentAmount / res.maxAmount.toDouble() * 100.0,
+            resourceType = res.type,
+            abilityName = abilityName
+        ))
     }
 
     fun subtractResource(amount: Int, type: Resource.Type, abilityName: String) {
-        if(resource.type == type) {
-            resource.subtract(amount)
+        var res = resource[type]
 
-            logEvent(Event(
-                eventType = Event.Type.RESOURCE_CHANGED,
-                amount = resource.currentAmount.toDouble(),
-                delta = -1 * amount.toDouble(),
-                amountPct = resource.currentAmount / resource.maxAmount.toDouble() * 100.0,
-                abilityName = abilityName
-            ))
-        } else {
-            logger.debug { "Attempted to subtract resource type $type but subject resource is ${resource.type}" }
+        if(res == null) {
+            logger.debug { "Attempted to subtract resource type $type that is not present for participant" }
+            return
         }
+
+        res.subtract(amount)
+
+        logEvent(Event(
+            eventType = Event.Type.RESOURCE_CHANGED,
+            amount = res.currentAmount.toDouble(),
+            delta = amount.toDouble(),
+            amountPct = res.currentAmount / res.maxAmount.toDouble() * 100.0,
+            resourceType = res.type,
+            abilityName = abilityName
+        ))
+    }
+
+    fun hasEnoughResource(type: Resource.Type, amount: Double) : Boolean {
+        if(amount == 0.0){return false} // for trinkets and other items that have MANA as the default
```

Comment:
Should this always return `true` if there is no resource cost?

Reply:
Oops, yes